src/client/dfs/dcache.c

```diff
@@ -904,48 +993,318 @@ drec_del_at_dact(dfs_dcache_t *dcache, dfs_obj_t *rec)
+	}
+
+	if (stbuf && !skip_stat) {
+		if (rec->dc_stated) {
+			memcpy(stbuf, &rec->dc_stbuf, sizeof(struct stat));
+		} else {
+			rc = entry_stat(dfs, dfs->th, parent->oh, rec->name, strlen(rec->name), rec,
+					true, stbuf, NULL);
+			if (rc != 0)
+				D_GOTO(out, rc);
+			memcpy(&rec->dc_stbuf, stbuf, sizeof(struct stat));
+			rec->dc_stated = true;
+			/* stat and dc_stated are updated, need to update the record in cache too.
+			 * the data buffer size is stored as an integer at the very beginning of
+			 * data.
+			 */
+			val_size = *((size_t *)value) & 0xFFFFFFFF;
+			rc       = dfs_obj_serialize(rec, (uint8_t *)value, &val_size);
+			if (rc != 0)
+				D_GOTO(out, rc);
+		}
+	}
+	if (mode && !skip_stat)
+		*mode = rec->mode;
+
+done:
+	*_rec = rec;
+out:
+	if (node_found)
+		shm_lru_node_dec_ref(node_found);
+	return rc;
 }
 
 int
 dcache_find_insert_rel(dfs_t *dfs, dfs_obj_t *parent, const char *name, size_t len, int flags,
-		       dfs_obj_t **rec, mode_t *mode, struct stat *stbuf)
+		       dfs_obj_t **_rec, mode_t *mode, struct stat *stbuf)
 {
+	dfs_obj_t      *rec;
+	char            key[DCACHE_KEY_PREF_SIZE + DFS_MAX_NAME];
+	size_t          key_len;
+	char           *value;
+	shm_lru_node_t *node_found = NULL;
+	int             rc;
+	size_t          val_size;
+	uint64_t       *ptr_key;
+
 	D_ASSERT(dfs->dcache != NULL);
 	D_ASSERT(name != NULL);
-	D_ASSERT(dfs->dcache->find_insert_rel_fn != NULL);
 
-	return dfs->dcache->find_insert_rel_fn(dfs->dcache, parent, name, len, flags, rec, mode,
-					       stbuf);
+	if (dfs->dcache->dd_type == DFS_CACHE_DRAM) {
+		D_ASSERT(dfs->dcache->dh.find_insert_rel_fn != NULL);
+		return dfs->dcache->dh.find_insert_rel_fn(dfs->dcache, parent, name, len, flags,
+							  _rec, mode, stbuf);
+	}
+
+	ptr_key    = (uint64_t *)key;
+	ptr_key[0] = dfs->dcache->shm.dd_pool_cont_hash;
+	if (parent) {
+		ptr_key[1] = parent->oid.lo;
+		ptr_key[2] = parent->oid.hi;
+	} else {
+		/* item under root dir */
+		ptr_key[1] = dfs->root.oid.lo;
+		ptr_key[2] = dfs->root.oid.hi;
+	}
+	memcpy(key + SHM_DCACHE_KEY_PREF_SIZE, name, len);
+	key_len = SHM_DCACHE_KEY_PREF_SIZE + len;
+	rc      = shm_lru_get(dfs->dcache->shm.d_shm_lru_dentry, key, key_len, &node_found,
+			      (void **)&value);
+	D_DEBUG(DB_TRACE, "dentry cache %s: name=" DF_PATH "\n", (rc != 0) ? "miss" : "hit",
+		DP_PATH(name));
+	if (rc == ENOENT) {
+		/* record is not found */
+		rc = dcache_add(dfs->dcache, parent, name, len, key, key_len, &rec, mode, stbuf);
+		if (rc)
+			D_GOTO(out, rc);
+	} else {
+		D_ALLOC_PTR(rec);
+		if (rec == NULL)
+			D_GOTO(out, rc = ENOMEM);
+
+		rc = dfs_obj_deserialize(dfs, flags, value, rec);
+		if (rc) {
+			D_FREE(rec);
+			D_GOTO(out, rc);
+		}
+
+		/** handle following symlinks outside of the dcache */
+		if (S_ISLNK(rec->mode) && !(flags & O_NOFOLLOW)) {
+			dfs_obj_t *sym;
+
+			rc = lookup_rel_path(dfs, parent, rec->value, flags, &sym, mode, stbuf, 0);
+			if (rc)
+				D_GOTO(out, rc);
+			rec = sym;
```

Comment:
From my understanding there is a memory leak here.
```suggestion
			D_FREE(rec);
			rec = sym;
```